### ISSUE TITLE
Use 2 maxpool2d with kernel = 3 as opposed to 1 maxpool2d with kernel = 5 when inferencing with CPU in SPPF layer

### ISSUE THREAD

**Issue by chutchatut** [ultralytics/ultralytics#9421]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

When no GPU is available switch from using 1 maxpool2d to using 2 consecutive maxpool2d to speed up inference.

### Use case

Deploying to environments without GPU.

### Additional

From my testing it seems to be 24.5% faster on CPU and 20% slower on GPU.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@chutchatut thanks for your suggestion and for sharing your findings on the SPPF layer's performance tweak for CPU inference! 🚀

Switching from a single `maxpool2d` with kernel size 5 to 2 consecutive `maxpool2d` layers with kernel size 3 for CPU inference, and observing a significant performance improvement, is quite interesting. Such adjustments are indeed crucial for deployments in GPU-constrained environments.

Before making any changes, we'll need to review the impact across various models and datasets to ensure it uniformly benefits or at least does not degrade accuracy/performance in most scenarios. Also, maintaining separate paths for GPU and CPU within the model might introduce some complexity, so we'll have to consider how to manage that effectively.

Your willingness to help by submitting a PR is greatly appreciated! Could you possibly share a bit more detail on your testing setup or even a brief code snippet of the changes? This will help us better understand the modifications and their effects. Something like:

```python
# Original
x = MaxPool2d(kernel_size=5, stride=1, padding=2)(x)

# Proposed change for CPU
x = MaxPool2d(kernel_size=3, stride=1, padding=1)(x)
x = MaxPool2d(kernel_size=3, stride=1, padding=1)(x)
```

This can be a starting point for discussion and potentially a valuable contribution to the project. Thanks again for your initiative!

**Comment by chutchatut** [comment]
Here's my testing setup. Please note that the run time of each block also include the time to clone the data.
But theoretically it should be around 28% faster so it's already quite close.

```python3
import torch
from torch import nn

x = torch.rand((1, 1000, 32))
------------------------------
%%timeit
y1 = x.clone()
with torch.no_grad():
    for _ in range(2):
        y1 = nn1.forward(y1)
# 155 µs ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
------------------------------
%%timeit
y2 = x.clone()
with torch.no_grad():
    y2 = nn2.forward(y2)
# 190 µs ± 14.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by chutchatut** [comment]
I also tried changing the SPPF layer to use this technique but I didn't test this code yet.

```python3
class SPPFF(nn.Module):
    """Spatial Pyramid Pooling - Fast (SPPFF) layer. This is slower on GPU"""

    def __init__(self, c1, c2, k=5):
        """
        Initializes the SPPF layer with given input/output channels and kernel size.

        This module is equivalent to SPP(k=(5, 9, 13)).
        """
        super().__init__()
        c_ = c1 // 2  # hidden channels
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = Conv(c_ * 4, c2, 1, 1)
        assert (k % 2 == 1, 'This module can only be used with odd kernel size')
        self.repeat = k // 2
        self.m = nn.MaxPool2d(kernel_size=3, stride=1, padding=1)

    def forward(self, x):
        y = self.cv1(x)
        out = [y]
        for _ in range(3):
            for _ in self.repeat:
                y = self.m(y)
            out.append(y)
        return self.cv2(torch.cat(out, 1))
```

This is just an idea of how I would change it.

**Comment by chutchatut** [comment]
After further testing. It seems that the performance increase decrease significantly with lower resolution.

When testing using shape = 64, 20, 20, 512.
The results only show 6% increase from
73.3 ms ± 5.26 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
to 
78.4 ms ± 6.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each).

**Comment by chutchatut** [comment]
Maybe it makes more sense to ask PyTorch team to improve the performance of MaxPool2d when stride=1 since it will benefit everyone using PyTorch and make the code much cleaner.

**Comment by chutchatut** [comment]
Since I move the issue to PyTorch, this is no longer necessary.

**Comment by glenn-jocher** [comment]
@chutchatut thanks for the update! It's great to see the initiative being taken to address the performance at its core with the PyTorch team. This approach will certainly help a broader audience benefit from any improvements made. If there's anything else we can assist with in the meantime, feel free to reach out. Happy coding! 🚀

**Comment by chutchatut** [comment]
@glenn-jocher Thanks for the offer. :smiley:

**Comment by glenn-jocher** [comment]
@chutchatut you're welcome! 😊 If you have any questions or need further assistance as you explore or contribute to the YOLOv8 project, don't hesitate to reach out. Happy coding!
